pkg/plugins/backendplugin/instrumentation/instrumentation.go

```diff
@@ -32,32 +35,38 @@ var (
 )
 
 const (
-	statusOK        = "ok"
-	statusError     = "error"
-	statusCancelled = "cancelled"
+	statusOK                      = "ok"
+	statusError                   = "error"
+	statusCancelled               = "cancelled"
+	httpStatusClientClosedRequest = 499
 )
 
 var logger = plog.New("plugin.instrumentation")
 
 // instrumentPluginRequest instruments success rate and latency of `fn`
 func instrumentPluginRequest(ctx context.Context, cfg Cfg, pluginCtx *backend.PluginContext, endpoint string, fn func() error) error {
 	status := statusOK
+	statusCode := http.StatusOK
 
 	start := time.Now()
 	timeBeforePluginRequest := log.TimeSinceStart(ctx, start)
 
 	err := fn()
 	if err != nil {
 		status = statusError
+		statusCode = http.StatusInternalServerError
 		if errors.Is(err, context.Canceled) {
 			status = statusCancelled
+			statusCode = httpStatusClientClosedRequest
 		}
 	}
 
 	elapsed := time.Since(start)
 	pluginRequestDuration.WithLabelValues(pluginCtx.PluginID, endpoint, string(cfg.Target)).Observe(float64(elapsed / time.Millisecond))
 	pluginRequestCounter.WithLabelValues(pluginCtx.PluginID, endpoint, status, string(cfg.Target)).Inc()
 
+	metrics.MPluginServiceRequestDurationSeconds.WithLabelValues(pluginCtx.PluginID, "grafana-backend", "datasource", endpoint, strconv.Itoa(statusCode)).Observe(float64(elapsed / time.Second))
```

Comment:
Maybe I'm missing something, but how come you can't achieve what you want with the existing `plugin_request_duration_milliseconds` metric?

Reply:
Is it ok for me to add and remove existing labels on `plugin_request_duration_milliseconds` ? To change the behavior of an existing label (`status`)? To change the list of bucket?

If yes to all of that then yes, I could just reuse the `plugin_request_duration_milliseconds` metric name.

Also I could define the metric in the `pkg/plugins/backendplugin/instrumentation/instrumentation.go`, like `plugin_request_duration_milliseconds`. It should work the same.

What do you advice to move forward?

Reply:
What's the motivation to introduce or change this behaviour in general? Is it for consistency with other Grafana metrics? We have plugin metrics in other places that follow this pattern so we would also need to update it there (if it made sense).

>To change the behavior of an existing label (status)? To change the list of bucket?

Do we need to do that when the existing one only uses 2 extra buckets? Considering these are part of the public API, I would be inclined not to modify unless really necessary (and then only do so in a minor/major release). 

Regarding statuses, HTTP codes don't always make sense here as a plugin may not be using HTTP to communicate downstream. Furthermore, the `app` label is hardcoded so does that provide much value?

Reply:
> What's the motivation to introduce or change this behaviour in general?

to improve plugin instrumentation

> Is it for consistency with other Grafana metrics? We have plugin metrics in other places that follow this pattern so we would also need to update it there (if it made sense).

Which ones? I might not be aware of those other plugin metrics but interested to learn more about them

> Do we need to do that when the existing one only uses 2 extra buckets?

Considering that one is in milliseconds and the other seconds, those are actually quite different buckets.
The current bucket list is not suited for the Prometheus plugin, which is by far the most used one.
Actually around 25% of requests have a longer duration than the highest bucket of the list (100ms). Which makes comonly used quantile like 90th, 95th, 99th, impossible to compute, as it quickly saturate above 70/80th.

> Considering these are part of the public API, I would be inclined not to modify unless really necessary (and then only do so in a minor/major release).

This is the reason I went for a completely different metric, so we don't have to think about that now. It is still experimental and in active development, we might experiment with multiple iterations in the following weeks.

> Furthermore, the app label is hardcoded so does that provide much value?

There will be measurements coming from the frontend too. I could create a separate metric for frontend and then merge them both in a new serie with a recording rule, but it would be simpler for me to just start with 1 metric with a label indicating where the measurement are coming from

Reply:
If you still think that this file is not suited for this kind of changes I can also bring everything inside our Prometheus plugins. Putting a new metric here would allow me to also check the data on different plugins, which could be very helpful for my project

Reply:
I see your point about the buckets 👍 How about we add a `plugin_request_duration_seconds` here and follow the same style as the existing microseconds metric? In time we could probably deprecate the older one (WDYT @bergquist?)

You can add a `source` label which indicates where you're measuring from (or in a follow up once you have a frontend metric) - BTW how will the frontend measurements work with this?

Reply:
Is there a reason why the histogram metric doesn't have a "status" label, and is excluding errors?
And then errors are monitored but on a separate metric
Ideally we only need one histogram metric, the `.._count` would be enough